File: cpp/scheduler.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <iomanip>
#include <sstream>

struct Interval {
    int start;
    int end;
    
    Interval(int start, int end) : start(start), end(end) {}
    
    bool overlaps(const Interval& other) const {
        return start < other.end && end > other.start;
    }
    
    std::string to_string() const {
        std::ostringstream oss;
        oss << "[" << std::setfill('0') << std::setw(2) << start 
            << ":00-" << std::setw(2) << end << ":00]";
        return oss.str();
    }
};

class Scheduler {
private:
    std::vector<Interval> intervals;
    int day_start;
    int day_end;
    
public:
    Scheduler(int day_start = 9, int day_end = 17) 
        : day_start(day_start), day_end(day_end) {}
    
    bool add(int start, int end) {
        Interval new_interval(start, end);
        
        if (new_interval.start >= new_interval.end) {
            return false;
        }
        
        for (const auto& existing : intervals) {
            if (new_interval.overlaps(existing)) {
                return false;
            }
        }
        
        intervals.push_back(new_interval);
        std::sort(intervals.begin(), intervals.end(),
                  [](const Interval& a, const Interval& b) {
                      return a.start < b.start;
                  });
        return true;
    }
    
    bool remove(int start, int end) {
        Interval target(start, end);
        for (auto it = intervals.begin(); it != intervals.end(); ++it) {
            if (it->start == target.start && it->end == target.end) {
                intervals.erase(it);
                return true;
            }
        }
        return false;
    }
    
    std::vector<Interval> free_slots(int min_duration = 1) {
        std::vector<Interval> free;
        int current = day_start;
        
        std::vector<Interval> sorted_intervals = intervals;
        std::sort(sorted_intervals.begin(), sorted_intervals.end(),
                  [](const Interval& a, const Interval& b) {
                      return a.start < b.start;
                  });
        
        for (const auto& interval : sorted_intervals) {
            if (current < interval.start) {
                if (interval.start - current >= min_duration) {
                    free.push_back(Interval(current, interval.start));
                }
            }
            current = std::max(current, interval.end);
        }
        
        if (current < day_end) {
            if (day_end - current >= min_duration) {
                free.push_back(Interval(current, day_end));
            }
        }
        
        return free;
    }
    
    std::vector<Interval> list_all() {
        std::vector<Interval> sorted_intervals = intervals;
        std::sort(sorted_intervals.begin(), sorted_intervals.end(),
                  [](const Interval& a, const Interval& b) {
                      return a.start < b.start;
                  });
        return sorted_intervals;
    }
    
    std::vector<Interval> find_conflicts(int start, int end) {
        Interval candidate(start, end);
        std::vector<Interval> conflicts;
        
        for (const auto& iv : intervals) {
            if (candidate.overlaps(iv)) {
                conflicts.push_back(iv);
            }
        }
        
        return conflicts;
    }
};

int main() {
    Scheduler scheduler(9, 17);
    
    std::cout << "Adding meetings..." << std::endl;
    scheduler.add(10, 11);
    scheduler.add(13, 14);
    scheduler.add(15, 16);
    
    std::cout << "\nScheduled intervals:" << std::endl;
    for (const auto& interval : scheduler.list_all()) {
        std::cout << "  " << interval.to_string() << std::endl;
    }
    
    std::cout << "\nFree slots:" << std::endl;
    for (const auto& slot : scheduler.free_slots()) {
        std::cout << "  " << slot.to_string() << std::endl;
    }
    
    std::cout << "\nTrying to add overlapping meeting (10:30-11:30):" << std::endl;
    if (scheduler.add(10, 11)) {
        std::cout << "  Added successfully" << std::endl;
    } else {
        std::cout << "  Failed - conflicts with existing meeting" << std::endl;
    }
    
    std::cout << "\nConflicts for 10:30-11:30:" << std::endl;
    auto conflicts = scheduler.find_conflicts(10, 11);
    for (const auto& conflict : conflicts) {
        std::cout << "  " << conflict.to_string() << std::endl;
    }
    
    return 0;
}

File: python/scheduler.py
#!/usr/bin/env python3
"""
Minimal scheduler for managing time intervals and finding free slots.
"""

from dataclasses import dataclass
from typing import List, Tuple

@dataclass
class Interval:
    """Time interval with start and end times."""
    start: int
    end: int
    
    def overlaps(self, other: 'Interval') -> bool:
        """Check if this interval overlaps with another."""
        return self.start < other.end and self.end > other.start
    
    def __str__(self):
        return f"[{self.start:02d}:00-{self.end:02d}:00]"

class Scheduler:
    """Minimal scheduler for managing intervals."""
    
    def __init__(self, day_start: int = 9, day_end: int = 17):
        self.intervals: List[Interval] = []
        self.day_start = day_start
        self.day_end = day_end
    
    def add(self, start: int, end: int) -> bool:
        """Add interval if it doesn't overlap with existing ones."""
        new_interval = Interval(start, end)
        
        if new_interval.start >= new_interval.end:
            return False
        
        for existing in self.intervals:
            if new_interval.overlaps(existing):
                return False
        
        self.intervals.append(new_interval)
        self.intervals.sort(key=lambda x: x.start)
        return True
    
    def remove(self, start: int, end: int) -> bool:
        """Remove matching interval."""
        target = Interval(start, end)
        for i, interval in enumerate(self.intervals):
            if interval.start == target.start and interval.end == target.end:
                self.intervals.pop(i)
                return True
        return False
    
    def free_slots(self, min_duration: int = 1) -> List[Interval]:
        """Find free time slots between scheduled intervals."""
        free = []
        current = self.day_start
        
        for interval in sorted(self.intervals, key=lambda x: x.start):
            if current < interval.start:
                if interval.start - current >= min_duration:
                    free.append(Interval(current, interval.start))
            current = max(current, interval.end)
        
        if current < self.day_end:
            if self.day_end - current >= min_duration:
                free.append(Interval(current, self.day_end))
        
        return free
    
    def list_all(self) -> List[Interval]:
        """List all scheduled intervals."""
        return sorted(self.intervals, key=lambda x: x.start)
    
    def find_conflicts(self, start: int, end: int) -> List[Interval]:
        """Find intervals that conflict with given time."""
        candidate = Interval(start, end)
        return [iv for iv in self.intervals if candidate.overlaps(iv)]

def main():
    """Example usage."""
    scheduler = Scheduler(day_start=9, day_end=17)
    
    print("Adding meetings...")
    scheduler.add(10, 11)
    scheduler.add(13, 14)
    scheduler.add(15, 16)
    
    print("\nScheduled intervals:")
    for interval in scheduler.list_all():
        print(f"  {interval}")
    
    print("\nFree slots:")
    for slot in scheduler.free_slots():
        print(f"  {slot}")
    
    print("\nTrying to add overlapping meeting (10:30-11:30):")
    if scheduler.add(10, 11):
        print("  Added successfully")
    else:
        print("  Failed - conflicts with existing meeting")
    
    print("\nConflicts for 10:30-11:30:")
    conflicts = scheduler.find_conflicts(10, 11)
    for conflict in conflicts:
        print(f"  {conflict}")

if __name__ == '__main__':
    main()
